### DISCUSSION TITLE
Inheritance results in empty field when type is of superclass

### DISCUSSION THREAD

**Discussion by larissaleite** [pydantic/pydantic#2851] [Question]
Hi! Thanks for the awesome work with pydantic 💪 

I have the following snippet:
```python
class Root(BaseModel):
    pass


class A(Root):
    a: bool

class B(Root):
    b: str

class MyModel(BaseModel):
    foo: str
    other: Optional[Root] = None


print(MyModel(**{"foo": "bar", "other": {"a": True}}))
```
This outputs `foo='bar' other=Root()`. If `A` is subclass of `Root`, is there a way to make `other` not empty?

**Comment by eddsalkield** [comment]
Setting the type of `other` to `Optional[Root]` means that the data will be validated against the model `Root`, which is what you see here.  `Root` has no way of knowing what its subclasses are.

If you intend that `other` instead be validated by class `A` or `B`, you should instead set its type to `Optional[Union[A, B]]`.  Then you get something like:

```python3
class Root(BaseModel):
    pass

class A(Root):
    a: bool

class B(Root):
    b: str
class MyModel(BaseModel):
    foo: str
    other: Optional[Union[A, B]] = None

print(MyModel(**{"foo": "bar", "other": {"a": True}}))
# other=A(a=True) foo='bar'
```

You may also find it useful to know that there's currently work being done on a [discriminated union](https://github.com/samuelcolvin/pydantic/issues/619) type, which would let pydantic select which of several types it's looking at based upon the value of a discriminator field.

Hope this helps!

**Reply by PrettyWood** [reply]
@eddsalkield is 100% right. If you don't want to right your `Union` manually you can register all subclasses of your `Root` class!
See https://github.com/samuelcolvin/pydantic/issues/2177#issuecomment-739578307 for an example
